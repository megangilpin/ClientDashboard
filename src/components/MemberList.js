import React from 'react';
import { makeStyles } from '@material-ui/core/styles';

import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import Typography from '@material-ui/core/Typography';
import Collapse from '@material-ui/core/Collapse';
import Divider from '@material-ui/core/Divider';
import ExpandLess from '@material-ui/icons/ExpandLess';
import ExpandMore from '@material-ui/icons/ExpandMore';
import AccountCircleIcon from '@material-ui/icons/AccountCircle';
import { formatDate } from '../utils';

const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    maxWidth: 360,
    backgroundColor: theme.palette.background.paper
  },
  nested: {
    paddingLeft: theme.spacing(4)
  },
  memberList: {
    display: 'flex',
    flexDirection: 'column'
  }
}));

function MemberList(props) {
  const classes = useStyles();
  const [open, setOpen] = React.useState(false);

  // Open/closes Member list dropdown
  const handleClick = () => {
    setOpen((preState) => !preState);
  };

  return (
    <List
      component="nav"
      aria-labelledby="nested-list-subheader"
      className={classes.root}
    >
      <ListItem button onClick={handleClick}>
        <ListItemIcon>
          <AccountCircleIcon />
        </ListItemIcon>
        <ListItemText primary="Members" />
        {open ? <ExpandLess /> : <ExpandMore />}
      </ListItem>
      <Collapse in={open} timeout="auto" unmountOnExit>
        <List component="div" disablePadding>
          {props.members.length > 0 ? (
            props.members.map((member) => (
              <React.Fragment key={member.id}>
                <ListItem
                  key={member.id}
                  className={classes.memberList}
                  alignItems="flex-start"
                >
                  <ListItemText primary={member.name} />
                  <Typography variant="caption" display="block">
                    {` ${member.title} - ${member.phone_number}`}
                  </Typography>
                  <Typography variant="caption" display="block">
                    {` Start Date - ${formatDate(member.created_at)}`}
                  </Typography>
                </ListItem>
                <Divider variant="inset" component="li" />
              </React.Fragment>
            ))
          ) : (
            <ListItem> No members exist</ListItem>
          )}
        </List>
      </Collapse>
    </List>
  );
}

export default MemberList;
