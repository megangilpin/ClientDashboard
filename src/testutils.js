const clients = [
  {
    id: '1',
    created_at: '2020-10-06T20:33:33.956Z',
    name: 'Client 1',
    headcount: 69,
    is_public: false,
    address_1: '714 Josefa Inlet',
    city: 'South Brendabury',
    zip_code: '69722-8987',
    state: 'MS'
  },
  {
    id: '2',
    created_at: '2020-10-06T20:33:33.956Z',
    name: 'Client 2',
    headcount: 8,
    is_public: false,
    address_1: '50 Broadway',
    city: 'New York',
    zip_code: '10031',
    state: 'NY'
  }
];

const members = [
  {
    created_at: '2020-08-11T16:36:27.612Z',
    id: '1',
    name: 'Destin Fahey',
    organization_id: 'organization_id 1',
    phone_number: '(840) 116-5157 x17522',
    title: 'Officer'
  },

  {
    created_at: '2020-03-09T03:38:36.139Z',
    id: '2',
    name: 'Miss Laverne Effertz',
    organization_id: 'organization_id 2',
    phone_number: '941-079-5733 x931',
    title: 'Executive'
  },

  {
    created_at: '2020-07-28T12:46:25.612Z',
    id: '3',
    name: 'Herminia Pfeffer',
    organization_id: 'organization_id 3',
    phone_number: '512.362.0270 x8191',
    title: 'Technician'
  }
];

const filteredMembers = [
  {
    created_at: '2020-08-11T16:36:27.612Z',
    id: '1',
    name: 'Destin Fahey',
    organization_id: 'organization_id 1',
    phone_number: '(840) 116-5157 x17522',
    title: 'Officer'
  }
];

export { clients, members, filteredMembers };
