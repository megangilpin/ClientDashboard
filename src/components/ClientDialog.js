import React from 'react';
import Organization from '../api/OrganizationService';
import Member from '../api/MembersService';
import { withStyles, makeStyles } from '@material-ui/core/styles';
import Dialog from '@material-ui/core/Dialog';
import MuiDialogTitle from '@material-ui/core/DialogTitle';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import { formatDate, filterMembers } from '../utils';

import MemberList from './MemberList';

const styles = (theme) => ({
  root: {
    margin: 0,
    padding: theme.spacing(2),
    backgroundColor: theme.palette.primary.light,
    color: theme.palette.common.white
  },
  closeButton: {
    position: 'absolute',
    right: theme.spacing(1),
    top: theme.spacing(1),
    color: theme.palette.common.white
  }
});

const DialogTitle = withStyles(styles)((props) => {
  const { children, classes, onClose, ...other } = props;
  return (
    <MuiDialogTitle disableTypography className={classes.root} {...other}>
      <Typography variant="h6">{children}</Typography>
      {onClose ? (
        <IconButton
          aria-label="close"
          className={classes.closeButton}
          onClick={onClose}
        >
          <CloseIcon />
        </IconButton>
      ) : null}
    </MuiDialogTitle>
  );
});

const useStyles = makeStyles((theme) => ({
  root: {
    width: 600
  },
  container: {
    minHeight: 440,
    minWidth: 440,
    padding: '2rem'
  },
  head: {
    backgroundColor: theme.palette.main
  }
}));

function ClientDialog(props) {
  const dialogClasses = useStyles();

  const [client, setClient] = React.useState({});
  const [members, setMembers] = React.useState({});

  React.useEffect(() => {
    if (props.clientId !== null) {
      Promise.all([Organization.getClient(props.clientId), Member.getMembers()])
        .then(function (results) {
          const client = results[0].data;
          setClient({ ...client });
          const filteredMembers = filterMembers(
            props.clientId,
            results[1].data
          );
          setMembers([...filteredMembers]);
        })
        .catch((error) => {
          console.log(error);
        });
    }
  }, [props.clientId]);

  return (
    <div>
      <Dialog
        onClose={props.handleClose}
        aria-labelledby="customized-dialog-title"
        open={props.open}
        scroll="paper"
      >
        <DialogTitle id="customized-dialog-title" onClose={props.handleClose}>
          Client Details
        </DialogTitle>
        <div className={dialogClasses.container}>
          <Grid container spacing={3}>
            <Grid item xs={12}>
              <Typography variant="h6">{client.name}</Typography>
            </Grid>
            <Grid item xs={12}>
              <Typography color="textSecondary">Address</Typography>
              <Typography variant="subtitle2">{client.address_1}</Typography>
              <Typography variant="subtitle2">
                {client.city}, {client.state}, {client.zip_code}
              </Typography>
            </Grid>
            <Grid item xs={4}>
              <Typography color="textSecondary">Account Type</Typography>
              <Typography variant="subtitle2">
                {client.is_public ? 'Public' : 'Private'}
              </Typography>
            </Grid>
            <Grid item xs={4}>
              <Typography color="textSecondary">Start Date</Typography>
              <Typography variant="subtitle2">
                {formatDate(client.created_at)}
              </Typography>
            </Grid>
            <Grid item xs={4}>
              <Typography color="textSecondary">Head Count</Typography>
              <Typography variant="subtitle2">{client.headcount}</Typography>
            </Grid>
            <Grid item xs={12}>
              <MemberList members={members} />
            </Grid>
          </Grid>
        </div>
      </Dialog>
    </div>
  );
}

export default ClientDialog;
