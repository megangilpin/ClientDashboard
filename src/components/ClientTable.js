import React from 'react';
import { withStyles, makeStyles } from '@material-ui/core/styles';
import Checkbox from '@material-ui/core/Checkbox';
import Paper from '@material-ui/core/Paper';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TablePagination from '@material-ui/core/TablePagination';
import TableRow from '@material-ui/core/TableRow';

import ClientDialog from '../components/ClientDialog';

import theme from '../themes/theme';

const StyledTableCell = withStyles((theme) => ({
  head: {
    backgroundColor: theme.palette.common.black,
    color: theme.palette.common.white
  },
  body: {
    fontSize: 14
  }
}))(TableCell);

// Sets columns names for table
const columns = [
  { id: 'name', label: 'Name', minWidth: 50 },
  {
    id: 'headcount',
    label: 'Head Count',
    minWidth: 170,
    align: 'center',
    format: (value) => value.toLocaleString('en-US')
  },
  {
    id: 'account',
    label: 'Public',
    minWidth: 170,
    align: 'center',
    format: (value) => value.toLocaleString('en-US')
  }
];

const useStyles = makeStyles((theme) => ({
  root: {
    width: '75vw',
    maxWidth: '700px'
  },
  container: {
    maxHeight: 440
  },
  head: {
    backgroundColor: theme.palette.main
  }
}));

function ClientTable(props) {
  const classes = useStyles();
  const [page, setPage] = React.useState(0);
  const [rowsPerPage, setRowsPerPage] = React.useState(10);
  const [clientRows, setClientRows] = React.useState([]);
  const [clientDialogOpen, setClientDialogOpen] = React.useState(false);
  const [clientId, setClientId] = React.useState(null);

  const handleClientDialogClose = () => {
    setClientDialogOpen(false);
  };

  const handleChangePage = (event, newPage) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (event) => {
    setRowsPerPage(+event.target.value);
    setPage(0);
  };

  const handleClick = (event, id) => {
    setClientId(id);
    setClientDialogOpen(true);
  };

  const createData = (id, name, headcount, is_public) => {
    return { id, name, headcount, is_public };
  };

  const createRows = (clients) => {
    let rows = clients.map((client) =>
      createData(client.id, client.name, client.headcount, client.is_public)
    );
    setClientRows([...rows]);
  };

  React.useEffect(() => {
    createRows(props.clients);
  }, [props.clients]);

  return (
    <Paper className={classes.root}>
      <TableContainer className={classes.container}>
        <Table stickyHeader aria-label="sticky table">
          <TableHead>
            <TableRow>
              {columns.map((column) => (
                <StyledTableCell
                  key={column.id}
                  align={column.align}
                  style={{ minWidth: column.minWidth }}
                >
                  {column.label}
                </StyledTableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {clientRows.length > 0
              ? clientRows
                  .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                  .map((row) => {
                    return (
                      <TableRow
                        hover
                        role="checkbox"
                        tabIndex={-1}
                        key={row.id}
                        onClick={(event) => handleClick(event, row.id)}
                      >
                        <TableCell
                          component="th"
                          id={row.id}
                          scope="row"
                          align="left"
                        >
                          {row.name}
                        </TableCell>
                        <TableCell align="center">{row.headcount}</TableCell>
                        <TableCell align="center" padding="checkbox">
                          <Checkbox
                            checked={row.is_public}
                            inputProps={{ 'aria-labelledby': 'Public Account' }}
                          />
                        </TableCell>
                      </TableRow>
                    );
                  })
              : null}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={clientRows.length}
        rowsPerPage={rowsPerPage}
        page={page}
        onChangePage={handleChangePage}
        onChangeRowsPerPage={handleChangeRowsPerPage}
      />
      <ClientDialog
        open={clientDialogOpen}
        handleClose={handleClientDialogClose}
        clientId={clientId}
      />
    </Paper>
  );
}

export default ClientTable;
