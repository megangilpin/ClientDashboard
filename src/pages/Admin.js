import React from 'react';
import Organization from '../api/OrganizationService';
import { makeStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';

import ClientTable from '../components/ClientTable';

const useStyles = makeStyles((theme) => ({
  root: {
    flexGrow: 1
  },
  menuButton: {
    marginRight: theme.spacing(2)
  },
  title: {
    flexGrow: 1
  },
  // necessary for content to be below app bar
  toolbar: theme.mixins.toolbar,
  content: {
    flexGrow: 1,
    backgroundColor: theme.palette.background.default,
    marginTop: theme.spacing(8),
    padding: theme.spacing(3),
    display: 'flex',
    justifyContent: 'center'
  }
}));

const Admin = (props) => {
  const classes = useStyles();
  const [clients, setClient] = React.useState([]);

  React.useEffect(() => {
    Organization.getClients()
      .then((res) => {
        setClient([...res.data]);
      })
      .catch((error) => {
        console.log(error);
      });
  }, []);

  return (
    <div className={classes.root}>
      <AppBar position="static">
        <Toolbar>
          <Typography variant="h6" className={classes.title}>
            Admin - Client Dashboard
          </Typography>
        </Toolbar>
      </AppBar>
      <main className={classes.content}>
        <div className={classes.toolbar} />
        <ClientTable clients={clients} />
      </main>
    </div>
  );
};

export default Admin;
