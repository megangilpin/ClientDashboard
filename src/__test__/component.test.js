import Admin from '../pages/Admin';
import ClientTable from '../components/ClientTable';
import ClientDialog from '../components/ClientDialog';
import MemberList from '../components/MemberList';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Typography from '@material-ui/core/Typography';
import TableRow from '@material-ui/core/TableRow';
import ListItem from '@material-ui/core/ListItem';
import { filteredMembers, clients } from '../testutils';

describe('<Admin /> rendering', () => {
  it('should render a <Typography /> for Nav', () => {
    const wrapper = shallow(<Admin />);
    expect(wrapper.find(Typography)).toHaveLength(1);
  });
});

describe('<ClientTable /> rendering', () => {
  it('should render <TableRow /> for number of clients', () => {
    const wrapper = mount(<ClientTable clients={clients} />);
    expect(wrapper.find(TableRow)).toHaveLength(3);
  });
  it('should render correctly', () => {
    const tree = renderer.create(<ClientTable clients={clients} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<ClientDialog /> rendering', () => {
  it('should render <Typography />', () => {
    const wrapper = mount(<ClientDialog clientId={1} open={true} />);
    expect(wrapper.find(Typography)).toHaveLength(12);
  });
});

describe('<MemberList /> rendering', () => {
  it('should render <ListItem /> for number of members', () => {
    const wrapper = mount(<MemberList members={filteredMembers} />);
    expect(wrapper.find(ListItem)).toHaveLength(1);
  });
  it('should render correctly', () => {
    const tree = renderer
      .create(<MemberList members={filteredMembers} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
