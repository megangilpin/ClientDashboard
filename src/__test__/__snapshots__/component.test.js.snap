// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientTable /> rendering should render correctly 1`] = `
<div
  className="MuiPaper-root makeStyles-root-6 MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    className="MuiTableContainer-root makeStyles-container-7"
  >
    <table
      aria-label="sticky table"
      className="MuiTable-root MuiTable-stickyHeader"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-9 MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 50,
              }
            }
          >
            Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-9 MuiTableCell-alignCenter MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 170,
              }
            }
          >
            Head Count
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-9 MuiTableCell-alignCenter MuiTableCell-stickyHeader"
            scope="col"
            style={
              Object {
                "minWidth": 170,
              }
            }
          >
            Public
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
  <div
    className="MuiTablePagination-root"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        className="MuiTablePagination-spacer"
      />
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          10
        </div>
        <input
          aria-hidden={true}
          className="MuiSelect-nativeInput"
          onAnimationStart={[Function]}
          onChange={[Function]}
          tabIndex={-1}
          value={10}
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        0-0 of 0
      </p>
      <div
        className="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Previous page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Next page"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
  <div />
</div>
`;

exports[`<MemberList /> rendering should render correctly 1`] = `
<nav
  aria-labelledby="nested-list-subheader"
  className="MuiList-root makeStyles-root-20 MuiList-padding"
>
  <div
    aria-disabled={false}
    className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemIcon-root"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
        />
      </svg>
    </div>
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        Members
      </span>
    </div>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root"
    />
  </div>
</nav>
`;
